Mergeable contract calls in volatile events

lukasz-golebiewski opened the pull request:
## Overview
See https://github.com/rchain/rchain/pull/2578


### JIRA ticket:
<sup>_Create it if there isn't one already._</sup>



### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>



### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

KentShikama committed:
Modify conflicts so that volatile events do not conflict

KentShikama committed:
Add test such that volatile produce/consume pair doesn't conflict

KentShikama committed:
Change checkNoConflicts test API to take base, b1, b2

KentShikama committed:
Remove produce/consume only when entire comm is volatile

As reviewed by @ArturGajowy

KentShikama committed:
Extract common code between checkConflicts/checkNoConflicts

KentShikama committed:
Add tests for mirrored diamond dag for checkConflicts/Merge

KentShikama committed:
Refactor buildBlockAncestorChannels to use extractP/C/Comm

As reviewed by @odeac

Format

KentShikama committed:
Label existing volatile produce/consume conflict test cases

Remove redundant test form

KentShikama committed:
Add test cases starting with !X

KentShikama committed:
Add test cases starting with !4

KentShikama committed:
Use 0 inside body where possible

KentShikama committed:
Add remaining tests for mergeability

ArturGajowy committed:
Parallelize the mergability test

ArturGajowy committed:
Make all mergability tests use the same channels

Using two different channels for testing is cheating,
as the merge then follows trivially.

ArturGajowy committed:
Add better assertion messages for mergability tests

ArturGajowy committed:
Move `base` parameter to the last position to better align with notation

I mean the notation used in Mike's spreadsheet and the one we use to
describe mergability test cases.

ArturGajowy committed:
Only perform the checks at one node in mergability tests

ArturGajowy committed:
Pick symmetric variants of dag at random in mergability tests

ArturGajowy committed:
Fix tests

ArturGajowy committed:
Fix the 4! C! test cases

ArturGajowy committed:
Make test case descriptions values, form a well-aligning DSL

ArturGajowy committed:
Extract notation constants for test rholang building blocks

ArturGajowy committed:
Add `Rho` value type for combining rholang snippets

ArturGajowy committed:
Fix the remaining test cases

ArturGajowy committed:
Add `coveredBy` to the mergability tests DSL for better alignment

lukasz-golebiewski committed:
Rename persist to persistent

In rspace.trace.Event

lukasz-golebiewski committed:
Store the persistent flag along with the Consume event

In rspace.trace.Event

lukasz-golebiewski committed:
Add persistent flag to ConsumeEvent

fixup! Add persistent flag to ConsumeEvent

lukasz-golebiewski committed:
Persistent consumes in volatile COMM events should be conflicting

lukasz-golebiewski committed:
Replace Z with Nil

lukasz-golebiewski committed:
Add missing "4X (!!4)" case

lukasz-golebiewski committed:
Store the persistent flag along with Produce events

fixup! Store the persistent flag along with Produce events

lukasz-golebiewski committed:
Add persistent flag to ProduceEvent

lukasz-golebiewski committed:
Use produce.persistent in the EstimatorHelper

lukasz-golebiewski committed:
Rename persist => persistent

dzajkowski commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -53,33 +59,152 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+      line: sourcecode.Line
+  ) =
+    diamondConflictCheck(base, b1, b2, numberOfParentsForDiamondTip = 1) /*>>
+      diamondConflictCheck(base, b2, b1, numberOfParentsForDiamondTip = 1)*/
```
what's the point of these comments?

lukasz-golebiewski replied:
This is already fixed

dzajkowski commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -53,33 +59,152 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+      b1: Rho,
+      b2: Rho,
+      numberOfParentsForDiamondTip: Int
+  )(implicit file: sourcecode.File, line: sourcecode.Line): Effect[Unit] =
```
👍

dzajkowski commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -53,33 +59,152 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+      "C! C! same !" -> conflicts(C_, C_, S0),
+      "C! C! diff !" -> conflictsForNow(C0, C1, S0 | S1)
+      // 4!! / !!4 row is similar to !4 / 4! and thus skipped
+      // C!! / !!C row is similar to !C / C! and thus skipped
```
is there a reason to not test this? or marking it with `coveredBy("!X (4!!)")`?

ArturGajowy replied:
The only reason is the amount of cases to test.
I originally asked for 36 cases, for everything in Mike's sheet except for pairs containing `!!` or peek. This though grows by the following factors:

 - whenever there's "match depending whether its same linear receive/send" we get one more case (same / diff)
 - whenever there's a comm event in the case, we need to test it locally to test the actual topic of this pr - volatile events. So eg. `4X !!4` spawns `4X (!!4)`, and `4! 4!` spawns `4! (4!)` and `(4!) (4!)`

BTW @lukasz-golebiewski I think we still need to add the "spawned" cases

dzajkowski commented on casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala:
```diff
@@ -86,14 +101,46 @@ object EstimatorHelper {
+          case CommEvent(Some(consume: ConsumeEvent), produces) =>
+            consume.channelsHashes ++ produces.map(_.channelsHash)
+          case e @ CommEvent(None, _) =>
+            throw new RuntimeException(s"Unexpected comm event $e")
```
raiseError will not work?

lukasz-golebiewski replied:
I changed the approach and this is no longer needed

dzajkowski reviewed (approved):
some small questions

ArturGajowy reviewed (changes requested):
My comments from Kent's PR still apply, I'll do the review of new changes (which seem to look good!) tomorrow. Please address the aforementioned comments though.

ArturGajowy commented:
Also, may I ask you to edit my `WIP pick symmetric variants of dag at random in mergability tests` commit to do what its title says (and no longer be WIP)? Basically it should switch places of the arguments b1 and b2 at random.

dzajkowski commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -53,33 +59,152 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+    //FIXME add missing cases for in-deploy COMM-s wherever there's a pair without X (a COMM)
+    //FIXME all `conflictsForNow` should eventually be replaced with `merges`
+    Map(
+      "!X !X"        -> conflictsForNow(S0, S0, Nil),
```
it would be so cool to have an additional field with reference to the reason why it conflicts for now; ticket, page, discord discussion, anything that would allow someone not familiar with the (large) problem space to catch up and see what needs to be done to enable this.

lukasz-golebiewski replied:
The JIRA ticket would be a better place than code

ArturGajowy replied:
I agree, although the general answer would be 'until we implement the corresponding merging rule / until we make it true'. In most cases, there are no longer objective justifications for them to merge, other than development time.

dzajkowski replied:
exactly my point. link to a spot that describes what needs to change so the bus factor is >1.

dzajkowski commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -53,33 +59,152 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+    Map(
+      "!X !X"        -> conflictsForNow(S0, S0, Nil),
+      "!X !4"        -> conflictsForNow(S0, S0, F_),
+      "!X (!4)"      -> merges(S0, S1 | F_, Nil),
```
I don't follow this case. why consume `S1` and not `S0`?

dzajkowski replied:
maybe to expand on my lack of understanding - it is confusing to see `"!X !4"        -> conflictsForNow(S0, S0, F_),` and `"!X (!4)"      -> merges(S0, S1 | F_, Nil),`. why the difference in data sent? does it matter?

lukasz-golebiewski replied:
This looks like an error, it should read `      "!X (!4)"      -> merges(S0, S0 | F_, Nil),`

lukasz-golebiewski replied:
Done

dzajkowski commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -53,34 +61,161 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+      "!X !X"        -> conflictsForNow(S0, S0, Nil),
+      "!X !4"        -> conflictsForNow(S0, S0, F_),
+      "!X (!4)"      -> merges(S0, S1 | F_, Nil),
+      "!X !C"        -> conflictsForNow(S0, S1, C1),
```
the document describing mergeability (TDDM?) comments here:
`If one produce got consumed and one didn't, they must be sending on different names`
this case sends data on the same name (`@0`)  so the result should be `!C !C`

is the assumption that this will be fixed somehow? I think that this isn't a possible state as `!X` will never happen with a `C` present on a common name. 

the next case I'm able to follow. there we introduce `C` on the right and it has to conflict.

dzajkowski replied:
missed the pattern match in C1

dzajkowski commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -53,34 +61,161 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+      "!X (!4)"      -> merges(S0, S1 | F_, Nil),
+      "!X !C"        -> conflictsForNow(S0, S1, C1),
+      "!X (!C)"      -> conflicts(S0, S1 | C_, Nil),
+      "!X 4X"        -> conflicts(S0, F_, Nil),
```
TDDM: `If we can prove they don't match, then they can merge.`

are there any plans to address this case? I guess this would mean running the SpatialMatcher.

lukasz-golebiewski replied:
I don't think running the SM will be necessary. No match === no COMM event and this is all the info we need and have

lukasz-golebiewski replied:
This one conflicts because `F_` and `S0` do match

dzajkowski replied:
that's the point: IF we can show that `!X` would not match with `F_` it's still fine to merge.

lukasz-golebiewski replied:
Right, so we need to add `"!X 4X - no match"        -> conflictsForNow(S0, F1, Nil),`

ArturGajowy replied:
I'm not sure if we should ever use the spatial matcher for refining those types of conflicts. We should reserve `conflictsForNow` for cases where it's clear that it should merge, and we can implement that quite easily.

ArturGajowy replied:
Having a `consider refining conflict detection by using spatial matcher` ticket is what we need here, I'd say

ArturGajowy replied:
Notice also that using spatial matcher requires having actual data and continuations, not just hashes. So it'd require a 'report-type' replay.

lukasz-golebiewski replied:
I added the `"!X 4X - no match" -> conflictsForNow(S0, F1, Nil),`. Making this and related cases merge is outside the scope of this PR

dzajkowski replied:
I was not arguing to solve it in this pr.
We have a description of how the merging rules should behave but we are ignoring half of the data there. who will remember to pick up that work? if it's part of this compiling documentation - we have a spot that tracks it.

ArturGajowy replied:
Just to be clear:

> I don't think running the SM will be necessary. No match === no COMM event and this is all the info we need and have

It will. Notice we're talking about merging two deploys (and their events) that weren't executing together. E.g.:

I deploy `@0!(0)`
You deploy `for (_ <- @0) { @1(1) }`

None of the deploys will have a COMM (or a match) in their log, but we can't merge them, as they'd leave TS in undefined state (more work is needed).

lukasz-golebiewski replied:
True

dzajkowski commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -53,34 +61,161 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+      "!X (C!)"      -> conflicts(S0, C_ | S0, Nil),
+      "!4 !4 same 4" -> conflicts(S0, S1, F_),
+      "!4 !4 diff 4" -> conflicts(S0, S1, F0 | F1),
+      "!4 (!4)"      -> merges(S0, S1 | F1, F0),
```
what about: `"!4 (!4)"      -> conflicts(S0, S1 | F0, F0),`?

dzajkowski replied:
it does not introduce new cases (and it should not conflict)

dzajkowski commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -53,34 +61,161 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+      "!4 !4 diff 4" -> conflicts(S0, S1, F0 | F1),
+      "!4 (!4)"      -> merges(S0, S1 | F1, F0),
+      "(!4) (!4)"    -> merges(S0 | F_, S0 | F_, Nil),
+      "!4 !C"        -> conflictsForNow(S0, S1, F0 | C1),
```
why is this marked as 'for now'? isn't this illegal?

dzajkowski replied:
seems its legal

lukasz-golebiewski committed:
Move away from using protobuf Events in EstimatorHelper

Use RSpace events in their stead

fixup! Move away from using protobuf Events in EstimatorHelper

fixup! Move away from using protobuf Events in EstimatorHelper

lukasz-golebiewski committed:
Extract the isVolatile function

lukasz-golebiewski committed:
Remove unnecessary pattern match

lukasz-golebiewski committed:
Leverage underscore

dzajkowski commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -53,34 +61,161 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+      "(!4) (!4)"    -> merges(S0 | F_, S0 | F_, Nil),
+      "!4 !C"        -> conflictsForNow(S0, S1, F0 | C1),
+      "!4 4X"        -> conflictsForNow(S0, F_, F_),
+      "!4 4!"        -> conflictsForNow(S0, F_, F0 | S1),
```
are we fine with having `F0` produce on `@0` as a result?

dzajkowski replied:
irrelevant

lukasz-golebiewski committed:
Remove extractors, use collect instead

lukasz-golebiewski committed:
Add "CX !!X" cases

dzajkowski commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -53,34 +61,163 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+      "!4 !!X"       -> conflictsForNow(S0, R1, F0),
+      "!4 !!4"       -> conflictsForNow(S0, R1, F_ | F1),
+      "!4 CX"        -> conflictsForNow(S0, C_, F_),
+      "!4 C!"        -> conflictsForNow(S0, C_, F0 | S1),
```
shouldn't this `C_` be `C1`?

ArturGajowy replied:
It could, but this makes the test more general. Notice there's no other send the C_ could match.

dzajkowski replied:
I was under the impression that we want to have a set of "rules" that describe how merging works.  the `!4 C!` has a very specific meaning. introducing generality in the implementation obfuscates the meaning. note that there is no generality in the term. it is a description of state.
IOW `!4 C!` is not the same as (S0, C_, F0 | S1). 
I understand that replacing the `_` will not yield equality but it will bring the implementation closer to what it should hold.

IF we had means to express the left hand in a better way than by discriminating with patterns - maybe it would be ok to try and write general laws. but we don't.

this is my 0.02$, do with it as you please.

ArturGajowy replied:
Notice we're talking about a test, not an implementation. The more general the tests, the more constrained the tested implementation.

Take for example `merges(S0, C_, F_)`.
Replacing it with `merges(S0, C0, F0)` poses the risk that the tested code will  only merge when the receives have the exact same matching pattern.

Granted, one could argue that `merges(S0, C_, F_)` will not spot cases where `merges(S0, C_, F0)` or `merges(S0, C0, F_)` would not hold. That would though require an odd explicit condition in the prod code, and can be fought by adding additional cases for specific data.

In general, I believe it's good to keep the test data minimal (C_ < C0) - which helps a better implementation, less over-specifying tests, and stays as close to the actual rules of the tested domain as possible.

dzajkowski commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -53,34 +61,163 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+      "!C !C"        -> conflictsForNow(S0, S0, C_),
+      "!C 4X"        -> conflictsForNow(S0, F_, C_),
+      "!C 4!"        -> conflictsForNow(S0, F_, C0 | S1),
+      "!C 4!!"       -> conflictsForNow(S0, F_, C0 | R1),
```
I don't agree with this usage of `_`. why not have it match exactly and remove any possibility for misunderstanding?

lukasz-golebiewski replied:
I agree, this is a good idea. It should make grokking the whole thing much easier

ArturGajowy replied:
I disagree :) This makes the tests more general and gives more chance for catching a bug when s/b modifies them wrongly / we break sth in code. We should only constrain the receives with concrete data when it's needed for the test to produce the events we need. Otherwise we should stick with the general `F_/C_` case.

lukasz-golebiewski replied:
I would argue that it's better to remove ambiguity during execution and split all cases containing `F_` into containing `F0` and `F1`. In any case let's not do that in this PR

dzajkowski commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -53,34 +61,163 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+      "!C 4!"        -> conflictsForNow(S0, F_, C0 | S1),
+      "!C 4!!"       -> conflictsForNow(S0, F_, C0 | R1),
+      "!C !!X"       -> conflictsForNow(S0, R1, C0),
+      "!C !!4"       -> conflictsForNow(S0, R1, C0 | F1),
```
I think that it would be beneficial to have a `infinite` or `impossible` marker that could hold e.g. `4X C!!` for the sake of completeness.

lukasz-golebiewski replied:
Shouldn't this `conflictForNow` and eventually `merge`? The impossible case is going to run out of phlo and so the deploy is going to be reverted without touching the tuplespace. So there shouldn't be any conflicts. Correct me if I'm wrong.

dzajkowski replied:
depends on what the interpreter will do :D

ArturGajowy replied:
`infinite/impossible` are just `merges`. Maybe it'd make sense to add a 'cause' enum to merges, so e.g.

```scala
val INFINITE = "Infinite loop is an error. Mergeable with everything."
merges(S0, R0, C_, INFINITE)
```

I'd maybe leave it for a future PR though?

lukasz-golebiewski replied:
Future PR sounds good

dzajkowski commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -53,34 +61,163 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+      "4X 4X"        -> conflictsForNow(F_, F_, Nil),
+      "4X 4!"        -> conflicts(F_, F_, S0),
+      // Skipping 4X 4!! merges, 4X !!X may merge or not, 4X !!4 may merge or not
+      "4X CX"        -> conflicts(F_, C_, Nil),
```
shouldn't this be marked as "for now"?

lukasz-golebiewski replied:
Yes, you're right

ArturGajowy commented on casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala:
```diff
@@ -84,16 +88,35 @@ object EstimatorHelper {
+      consumeEvents = ancestorEvents.collect { case c: Consume => c }
+      commEvents    = ancestorEvents.collect { case c: COMM    => c }
+
+      producesInCommEvents = commEvents.map(_.produces).flatten
```
flatMap

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/EstimatorHelperTest.scala:
```diff
@@ -88,14 +88,14 @@ class EstimatorHelperTest
-  it should "conflict if their deploys contain same channel in deployLog" in withStorage {
+  ignore should "conflict if their deploys contain same channel in deployLog" in withStorage {
     implicit blockStore => implicit blockDagStorage =>
       testConflict[Task] { deploy =>
```
these two tests need to be fixed. Their names should be:

`detect conflicts within deployLog-s`
`detect conflicts within paymentLog-s`

there should also be a third one:

`detect conflicts between deployLog and paymentLog`

The bodies should be virtually the same, except `testConflict` will now take two deploy modifying functions: one for each deploy.

first deploy should do a produce on channel A
second deploy should do a consume on channel A
(this conflicts)

the only difference between those three tests is which logs contain the above events.

Ideally, we'd make the 3 tests say:

```scala
val deployLogSetter = deploy => log => deploy.copy(deployLog = log)
val paymentLogSetter = deploy => log => deploy.copy(paymnentLog = log)
testConflicts(deployLogSetter, deployLogSetter)
testConflicts(paymentLogSetter, paymentLogSetter)
testConflicts(deployLogSetter, paymentLogSetter)
```
//lenses ftw!

lukasz-golebiewski replied:
```
first deploy should do a produce on channel A
second deploy should do a consume on channel A
(this conflicts)
```
This is a huge oversimplification as this isn't always going to conflict because the produce and consume might not match. Hence I would argue that this test should be removed given how fine grained the tests in `MultiParentCasperMergeSpec` are going to be

ArturGajowy replied:
> the produce and consume might not match
remember we're in control of the code used for doing the deploy. The `testConfclits` funciton ensures they do match, by not specifying a pattern on the for.

MergeSpec can't test inclusion of payment logs until they're actually used, and is far heavier than this is. If it's such a pain - leave those ignored, I'll fix it when I come to it.

lukasz-golebiewski replied:
Nah, I'll unignore them

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -53,34 +61,163 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+      b2: Rho,
+      numberOfParentsForDiamondTip: Int
+  )(implicit file: sourcecode.File, line: sourcecode.Line): Effect[Unit] =
+    Debug.print[Effect](base, b1, b2, numberOfParentsForDiamondTip) >>
```
please remove this, ideally by editing the commit that introduced it. Sorry about that.

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -53,34 +61,163 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
-        _      <- nodes(1).receive()
-        _      <- nodes(0).receive()
-        _      <- nodes(1).receive()
+    def echoContract(no: Int) =
```
maybe un-ignore the above test and use `conflictsForNow`?

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -123,7 +260,7 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
     val registryRho =
       """
-        |// Expected output
+        |"Expected output".
```
Why this change? Doesn't it break the test?

lukasz-golebiewski replied:
No idea, I didn't author it. Will revert though as it doesn't look sensible

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -53,34 +61,163 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+      // C!! / !!C row is similar to !C / C! and thus skipped
+      "CX !!X no match" -> conflictsForNow(R0, C1, Nil),
+      "CX !!X match"    -> conflicts(R0, C0, Nil)
+    ).values.toList.parSequence_
```
We should verify that we have cases generated by the following code,
which amounts to 69 cases (not including same/diff distinction).

We'd check that by:

having the Map defining the test cases named `tests`
asserting that all cases `c` defined in `withLocals` have a key in the `tests` map that starts with `c`

```

val all = List("!X", "!4", "!C", "4X", "4!", "4!!", "PX", "P!", "P!!", "!!X", "!!4", "!!C", "CX", "C!", "C!!")
val considered = all.filterNot(_.contains("P")).filterNot(_.contains("!!"))

val pairs = considered.combinations(2)
val diagonal = considered.map(x => List(x, x))
val cases = (pairs ++ diagonal).toList

def isComm(s: String) = !s.contains("X")
def makeLocal(s: String): List[String] = isComm(s) match {
  case false => List(s)
  case true => List(s, s"($s)")
}

def makeLocals(v: List[String]): List[List[String]] = for {
  a <- makeLocal(v(0))
  b <- makeLocal(v(1))
} yield List(a, b)

val withLocals = cases.flatMap(makeLocals)

val s = withLocals.map(_.mkString(" ")).mkString("\n")
```

output:

```
!X !4
!X (!4)
!X !C
!X (!C)
!X 4X
!X 4!
!X (4!)
!X CX
!X C!
!X (C!)
!4 !C
!4 (!C)
(!4) !C
(!4) (!C)
!4 4X
(!4) 4X
!4 4!
!4 (4!)
(!4) 4!
(!4) (4!)
!4 CX
(!4) CX
!4 C!
!4 (C!)
(!4) C!
(!4) (C!)
!C 4X
(!C) 4X
!C 4!
!C (4!)
(!C) 4!
(!C) (4!)
!C CX
(!C) CX
!C C!
!C (C!)
(!C) C!
(!C) (C!)
4X 4!
4X (4!)
4X CX
4X C!
4X (C!)
4! CX
(4!) CX
4! C!
4! (C!)
(4!) C!
(4!) (C!)
CX C!
CX (C!)
!X !X
!4 !4
!4 (!4)
(!4) !4
(!4) (!4)
!C !C
!C (!C)
(!C) !C
(!C) (!C)
4X 4X
4! 4!
4! (4!)
(4!) 4!
(4!) (4!)
CX CX
C! C!
C! (C!)
(C!) C!
(C!) (C!)
```

lukasz-golebiewski replied:
Ok, in a follow up PR

ArturGajowy reviewed (changes requested):
We should change the name of this PR to 'Volatile events mergable", because that's the scope of mergability that it provides. We should also link to the relevant jira issue (not the 'mergable contract calls' one - or we need to split that one)

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -53,34 +61,163 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+      "!X !X"        -> conflictsForNow(S0, S0, Nil),
+      "!X !4"        -> conflictsForNow(S0, S0, F_),
+      "!X (!4)"      -> merges(S0, S0 | F_, Nil),
+      "!X !C"        -> conflictsForNow(S0, S1, C1),
```
This should say
```
"!X !C" -> conflictsForNow(S0, S0, C_)
```

lukasz-golebiewski replied:
Why?
Besides, `S0, S0, C_` will always conflict

ArturGajowy replied:
yeah, sorry - I was wrong about this one. It's correct.
(`"!X !C"        -> conflictsForNow(S0, S1, C1),`)

ArturGajowy committed:
Add descriptions for test cases, fix/refine some of them

Cases which changed beyond just adding description:
!X !4
!X (!C)
!4 !4 diff 4
!4 (!4)
4X 4!
4! C! same !
CX !!X (removed match/no match distinction till we decide to use spatial matcher)

ArturGajowy commented:
I think I've managed to classify TDDM's cases into 4 types, that will allow for clean implementation of the underlying merging rules.

It also occurs to me, that there are really just 2 merge criteria:
- the things merged cannot cause additional comms (can't match each other - we need to find them matches that prevent that in the original deploys)
- the things merged cannot be consuming the same linear event

The split that occured to me is:
- whether the incoming events have same or different polarity (send/receive).
  same polarity cant produce additional comms
  different polarity cant match the same linear event
- within same/diff polarity category - whether the considered case could lead to the wrong specific to its polarity category

See the description of cases:
SAME_POLARITY_MERGE
COULD_MATCH_SAME
HAD_ITS_MATCH
INCOMING_COULD_MATCH

lukasz-golebiewski committed:
Unignore testing conflicts on stdout

lukasz-golebiewski committed:
conflictsForNow preserves place of failure

lukasz-golebiewski committed:
Improve error message of Blake2b256Hash requirement

lukasz-golebiewski commented:
bors r+
